attestation-agent/attester/src/tsm_report/mod.rs

```diff
@@ -0,0 +1,157 @@
+pub struct TsmReportPath {
+    path: PathBuf,
+}
+
+impl Drop for TsmReportPath {
+    fn drop(&mut self) {
+        let _ = std::fs::remove_dir(self.path.as_path())
+            .map_err(|e| log::error!("Failed to remove TSM Report directory: {}", e));
+    }
+}
+
+impl TsmReportPath {
+    pub fn new(wanted: TsmReportProvider) -> Result<Self, TsmReportError> {
+        if !Path::new(TSM_REPORT_PATH).exists() {
+            return Err(TsmReportError::NoTsmReports);
+        }
+
+        let p = tempdir_in(TSM_REPORT_PATH).map_err(TsmReportError::Open)?;
+
+        // Remove the Drop set by tempdir_in() since it errors on ConfigFS
+        // and leaks the created path. We implement our own Drop that removes the
+        // path (rmdir way) when TsmReportPath instance goes out of scope.
+        let path = p.into_path();
+
+        check_tsm_report_provider(path.as_path(), wanted).map_err(|e| {
+            let _ = std::fs::remove_dir(path.as_path());
+            e
+        })?;
+
+        Ok(Self { path })
+    }
+    pub fn attestation_report(
+        &self,
+        provider_data: TsmReportData,
+    ) -> Result<Vec<u8>, TsmReportError> {
+        let report_path = self.path.as_path();
+
+        let report_data = match provider_data {
+            TsmReportData::Tdx(inblob) => inblob,
+            TsmReportData::Sev(privlevel, inblob) => {
+                // TODO: untested
+                std::fs::write(report_path.join("privlevel"), vec![privlevel])
+                    .map_err(|e| TsmReportError::Access("privlevel", e))?;
+                inblob
+            }
+        };
+
+        if report_data.is_empty() {
+            return Err(TsmReportError::InblobLen);
+        }
+
+        std::fs::write(report_path.join("inblob"), report_data)
+            .map_err(|e| TsmReportError::Access("inblob", e))?;
+
+        let q = std::fs::read(report_path.join("outblob"))
+            .map_err(|e| TsmReportError::Access("outblob", e))?;
+
+        check_inblob_write_race(report_path)?;
+
+        Ok(q)
+    }
+    pub fn supplemental_data(&self) -> Result<Vec<u8>, TsmReportError> {
+        let report_path = self.path.as_path();
+
+        let aux = std::fs::read(report_path.join("auxblob"))
+            .map_err(|e| TsmReportError::Access("auxblob", e))?;
+
+        check_inblob_write_race(report_path)?;
+
+        Ok(aux)
+    }
+}
+
+// check_inblob_write_race checks that the returned outblob/auxblob
+// matches the quote generation request originally triggered when
+// inblob was written by the TsmReportPath instance. It prevents
+// the race condition that someone else could use the same temporary
+// directory to generate a quote.
+fn check_inblob_write_race(report_path: &std::path::Path) -> Result<(), TsmReportError> {
+    let g = std::fs::read_to_string(report_path.join("generation"))
+        .map_err(|e| TsmReportError::Access("generation", e))?;
+
+    let generation = g.trim_matches('\n').to_string().parse::<u32>().unwrap();
```

Comment:
do you consider using 'map_err' instead of 'unwrap'?

Reply:
fixed